tests/lib/c_lib/src/main.c

```diff
@@ -319,6 +322,181 @@ void test_bsearch(void)
+
+	zassert_true(isgraph(str1), "isgraph 1");
+	zassert_false(isgraph(str4), "isgraph ' '");
+
+	zassert_false(isprint(ch), "isprint 0x7f");
+	zassert_true(isprint(str1), "isprint '1'");
+
+	zassert_false(isspace(str2), "isspace 'a'");
+	zassert_true(isspace(str4), "isspace ' '");
+
+	zassert_false(isupper(str2), "isupper 'a'");
+	zassert_true(isupper(upper), "isupper 'A'");
+
+	zassert_true(isxdigit(str2), "isxdigit 'a'");
+	zassert_false(isxdigit(str5), "isxdigit 'z'");
+}
+
+/**
+ *
+ * @brief Test character operation
+ *
+ * @see memchr(), memcpy(), memmove().
+ *
+ */
+void test_memstr(void)
+{
+	const char str[] = "testfunction";
+	int ch1 = 'e', ch2 = 'z';
+	char arr[10], move_arr[6] = "12123";
+	const char num[6] = "12121";
+
+	zassert_not_null(memchr(str, ch1, strlen(str)),
+		     "memchr 'testfunction' 'e' ");
+	zassert_is_null(memchr(str, ch2, strlen(str)),
+		     "memchr 'testfunction' 'z'");
+
+	zassert_not_null(memcpy(arr, num, sizeof(num)), "memcpy return NULL");
+	zassert_equal(memcmp(num, arr, sizeof(num)), 0,
+		     "memcpy failed");
+
+	zassert_not_equal(memmove(move_arr + 2, move_arr, 3), NULL,
+		     "memmove error");
+	zassert_equal(memcmp(num, move_arr, 6), 0,
+		     "memmove failed");
+}
+
+/**
+ *
+ * @brief test str operate functions
+ *
+ * @see strcat(), strcspn(), strncat().
+ *
+ */
+void test_str_operate(void)
+{
+	char str1[10] = "aabbcc", ncat[10] = "ddee";
+	char *str2 = "b";
+	char *str3 = "d";
+	int ret;
+	char *ptr;
+
+	zassert_not_null(strcat(str1, str3), "strcat false");
+	zassert_equal(strcmp(str1, "aabbccd"), 0, "test strcat failed");
+
+	ret = strcspn(str1, str2);
+	zassert_equal(ret, 2, "strcspn failed");
+	ret = strcspn(str1, str3);
+	zassert_equal(ret, 6, "strcspn not found str");
+
+	zassert_true(strncat(ncat, str1, 2), "strncat failed");
+	zassert_equal(strcmp(ncat, "ddeeaa"), 0, "strncat failed");
+
+	zassert_is_null(strrchr(ncat, 'z'),
+		       "strrchr not found this word. failed");
+	ptr = strrchr(ncat, 'e');
+	zassert_equal(strcmp(ptr, "eaa"), 0, "strrchr failed");
+
+	zassert_is_null(strstr(str1, "xyz"), "strstr aabbccd with xyz failed");
+	zassert_not_null(strstr(str1, str2), "strstr aabbccd with b failed");
+}
+
+/**
+ *
+ * @brief test strtol function
+ *
+ */
+void test_strtoxl(void)
+{
+	char buf1[20] = "10379aegi";
+	char buf2[20] = "10379aegi";
+	char buf3[20] = "010379aegi";
+	char buf4[20] = "0x10379aegi";
+	char *stop = NULL;
+	long ret;
+	unsigned long retul;
+
+	ret = strtol(buf3, &stop, 8);
+	zassert_true((ret == 543), "strtol base = 8 failed");
+	ret = strtol(buf1, &stop, 10);
+	zassert_true((ret == 10379), "strtol base = 10 failed");
+	retul = strtoul(buf2, &stop, 10);
+	zassert_true((ret == 10379), "strtoul base = 10 failed");
+	ret = strtol(buf4, &stop, 16);
+	zassert_true((ret == 17004974), "strtol base = 16 failed");
+}
+
+/**
+ *
+ * @brief test convert function
+ *
+ */
+void test_tolower_toupper(void)
+{
+	int upper1 = 'A', upper2 = 'Z', lower1 = 'a', lower2 = 'z';
```

Comment:
Yes, I will change it by your comments.

Reply:
`toupper()` and `tolower()` just convert 1 character, so I test in this way. String can't be converted via `toupper()` and `tolower()`.

Reply:
I mean do something like this, so you test more than just letters.
```
  const char test[] = "aZ09zA\t\n#!";
  const char toup[] = "AZ09ZA\t\n#!";
  const char tolw[] = "az09za\t\n#!";
  char up[sizeof(test)];
  char lw[sizeof(test)];
  size_t idx = 0;

  while (test[idx] != 0) {
    up[idx] = toupper(test[idx]);
    lw[idx] = tolower(test[idx]);
    ++idx;
  }
  up[idx] = lw[idx] = 0;
  zassert_equal(strcmp(up, toup), 0);
  zassert_equal(strcmp(lw, toup), 0);
```
(I haven't run this so it may be wrong.)